mixnet/protocol/temporalmix.py

```diff
@@ -0,0 +1,145 @@
+import random
+from abc import abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import TypeVar
+
+from framework.framework import Framework, Queue
+
+
+class TemporalMixType(Enum):
+    NONE = "none"
+    PURE_COIN_FLIPPING = "pure-coin-flipping"
+    PURE_RANDOM_SAMPLING = "pure-random-sampling"
+    PERMUTED_COIN_FLIPPING = "permuted-coin-flipping"
+
+
+@dataclass
+class TemporalMixConfig:
+    mix_type: TemporalMixType
+    # The minimum size of queue to be mixed.
+    # If the queue size is less than this value, noise messages are added.
+    min_queue_size: int
+    # Generate the seeds used to create the RNG for each queue that will be created.
+    seed_generator: random.Random
+
+    def __post_init__(self):
+        assert self.seed_generator is not None
+        assert self.min_queue_size > 0
+
+
+T = TypeVar("T")
+
+
+class TemporalMix:
+    @staticmethod
+    def queue(
+        config: TemporalMixConfig, framework: Framework, noise_msg: T
+    ) -> Queue[T]:
+        match config.mix_type:
+            case TemporalMixType.NONE:
+                return NonMixQueue(framework, noise_msg)
+            case TemporalMixType.PURE_COIN_FLIPPING:
+                return PureCoinFlipppingQueue(
+                    config.min_queue_size,
+                    random.Random(config.seed_generator.random()),
+                    noise_msg,
+                )
+            case TemporalMixType.PURE_RANDOM_SAMPLING:
+                return PureRandomSamplingQueue(
+                    config.min_queue_size,
+                    random.Random(config.seed_generator.random()),
+                    noise_msg,
+                )
+            case TemporalMixType.PERMUTED_COIN_FLIPPING:
+                return PermutedCoinFlipppingQueue(
+                    config.min_queue_size,
+                    random.Random(config.seed_generator.random()),
+                    noise_msg,
+                )
+            case _:
+                raise ValueError(f"Unknown mix type: {config.mix_type}")
+
+
+class NonMixQueue(Queue[T]):
+    """
+    Queue without temporal mixing. Only have the noise generation when the queue is empty.
+    """
+
+    def __init__(self, framework: Framework, noise_msg: T):
+        self.__queue = framework.queue()
+        self.__noise_msg = noise_msg
+
+    async def put(self, data: T) -> None:
+        await self.__queue.put(data)
+
+    async def get(self) -> T:
+        if self.__queue.empty():
+            return self.__noise_msg
+        else:
+            return await self.__queue.get()
+
+    def empty(self) -> bool:
+        return self.__queue.empty()
+
+
+class MinSizeMixQueue(Queue[T]):
+    def __init__(self, min_pool_size: int, rng: random.Random, noise_msg: T):
+        super().__init__()
+        # Assuming that simulations run in a single thread
+        self._queue: list[T] = []
```

Comment:
I think that during the initialization we should fill the queue up with `noise_msg`? This way we should have the same initial state for all nodes. Then the real message is going to be added at the end of the queue. Otherwise, if we `put` before we `get`, then the real message is going to be in front of the queue which is going to impact the simulation.

Reply:
Good point. As we discussed in the chat, I will initialize the queue with `min_queue_size` number of messages.
(As we discussed, I will also run experiments with the current implementation to see the differences).